pkg/session/state.go

```diff
@@ -209,85 +206,125 @@ func (spt *SessionState) RemoveFromRetrieval(storageProviderId peer.ID, retrieva
-		status.suspensionDuration = spt.config.SuspensionDuration
-	}
-
-	if status.isSuspended() {
-		logger.Warnf(
-			"Suspending storage provider for %s after %d failures within %s: %s",
-			status.suspensionDuration,
-			len(status.failures),
-			time.Since(status.failures[0]),
-			storageProviderId,
-		)
+	// EMA of successes (0.0 & 1.0)
+	if !status.success.initialized {
+		status.success.initialized = true
+		status.success.value = current
+	} else {
+		status.success.value = (1-spt.config.SuccessAlpha)*current + spt.config.SuccessAlpha*status.success.value
 	}
-
-	// Write updated status back to map
 	spt.spm[storageProviderId] = status
-	return nil
 }
 
-func (spt *SessionState) RegisterConnectTime(storageProviderId peer.ID, current time.Duration) {
+func (spt *SessionState) RecordConnectTime(storageProviderId peer.ID, current time.Duration) {
 	spt.lk.Lock()
 	defer spt.lk.Unlock()
 
+	currentMs := uint64(current.Milliseconds())
+
 	status := spt.spm[storageProviderId]
 	// EMA of connect time
-	if !status.connectTimeInitialized {
-		status.connectTimeInitialized = true
-		status.connectTime = current
+	if !status.connectTimeMs.initialized {
+		status.connectTimeMs.initialized = true
+		status.connectTimeMs.value = currentMs
 	} else {
-		status.connectTime = time.Duration((1-spt.config.ConnectTimeAlpha)*float64(current) + spt.config.ConnectTimeAlpha*float64(status.connectTime))
+		status.connectTimeMs.value = uint64((1-spt.config.ConnectTimeAlpha)*float64(currentMs) + spt.config.ConnectTimeAlpha*float64(status.connectTimeMs.value))
 	}
 	spt.spm[storageProviderId] = status
+
+	if !spt.overallConnectTimeMs.initialized {
+		spt.overallConnectTimeMs.initialized = true
+		spt.overallConnectTimeMs.value = currentMs
+	} else {
+		spt.overallConnectTimeMs.value = uint64((1-spt.config.OverallConnectTimeAlpha)*float64(currentMs) + spt.config.OverallConnectTimeAlpha*float64(spt.overallConnectTimeMs.value))
+	}
 }
 
-func (spt *SessionState) CompareStorageProviders(protocol multicodec.Code, a, b peer.ID, mda, mdb metadata.Protocol) bool {
-	if protocol == multicodec.TransportGraphsyncFilecoinv1 {
-		gsmda := mda.(*metadata.GraphsyncFilecoinV1)
-		gsmdb := mdb.(*metadata.GraphsyncFilecoinV1)
-		// prioritize verified deals over not verified deals
-		if gsmda.VerifiedDeal != gsmdb.VerifiedDeal {
-			return gsmda.VerifiedDeal
-		}
+func (spt *SessionState) ChooseNextProvider(peers []peer.ID, mda []metadata.Protocol) int {
+	spt.lk.Lock()
+	defer spt.lk.Unlock()
 
-		// prioritize fast retrievel over not fast retrieval
-		if gsmda.FastRetrieval != gsmdb.FastRetrieval {
-			return gsmda.FastRetrieval
+	// score all peers, a float >=0 each, higher is better
+	scores := make([]float64, len(peers))
+	var tot float64
+	ind := make([]int, len(peers)) // index into peers that we can sort without changing peers
+	for ii, p := range peers {
+		ind[ii] = ii
+		gsmd, _ := mda[ii].(*metadata.GraphsyncFilecoinV1)
+		scores[ii] = spt.scoreProvider(p, gsmd)
```

Comment:
just something to consider: maybe at some point we start squaring these values to reduce randomness. (or perhaps make that it's own adjustable parameter)

Reply:
I'm hoping we can come up with ways to be able to know what to do here without guessing too much; I'm trying modelling for now, but insights and A/B testing live might be better.